crates/wasm-encoder/src/component/types.rs

```diff
@@ -530,21 +530,10 @@ impl Encode for PrimitiveValType {
 #[cfg(feature = "wasmparser")]
 impl From<wasmparser::PrimitiveValType> for PrimitiveValType {
     fn from(ty: wasmparser::PrimitiveValType) -> Self {
-        match ty {
-            wasmparser::PrimitiveValType::Bool => PrimitiveValType::Bool,
-            wasmparser::PrimitiveValType::S8 => PrimitiveValType::S8,
-            wasmparser::PrimitiveValType::U8 => PrimitiveValType::U8,
-            wasmparser::PrimitiveValType::S16 => PrimitiveValType::S16,
-            wasmparser::PrimitiveValType::U16 => PrimitiveValType::U16,
-            wasmparser::PrimitiveValType::S32 => PrimitiveValType::S32,
-            wasmparser::PrimitiveValType::U32 => PrimitiveValType::U32,
-            wasmparser::PrimitiveValType::S64 => PrimitiveValType::S64,
-            wasmparser::PrimitiveValType::U64 => PrimitiveValType::U64,
-            wasmparser::PrimitiveValType::F32 => PrimitiveValType::F32,
-            wasmparser::PrimitiveValType::F64 => PrimitiveValType::F64,
-            wasmparser::PrimitiveValType::Char => PrimitiveValType::Char,
-            wasmparser::PrimitiveValType::String => PrimitiveValType::String,
-        }
+        crate::reencode::utils::component_primitive_val_type(
+            &mut crate::reencode::RoundtripReencoder,
+            ty,
+        )
```

Comment:
I feel like I had ambitions for the style of implementation in `wasm-mutate` which may not have quite panned out. Specifically having all the default trait methods call helper functions as opposed to putting the implementation in the trait method itself.

Could the `utils` module instead be not used that much ideally? For example could this be `RoundtripReencoder.component_primitive_val_type(ty)`?

Reply:
Actually thinking a bit more on this, what do you think about removing the `utils` module entirely? Instead having only defaulted trait methods?

Reply:
My reasoning for having the default dispatch implementations as free functions is that there may be cases where one wants to override one of the methods but just to inspect something and not to change the actual functionality. Instead of then needing to copy-past in the default impl, one can simply call the utils function.

I agree that most of these methods will likely not need that, and we could remove them if we think that the default impl is short enough. My prime usecase would be the reencoding of instructions - if I just need to change a few match arms, I don't want to copy-paste the entire thing (especially since I do want to preserve all recursive reencoding behaviour of indices).

Reply:
Ok I think that's reasonable yeah 👍